onnxruntime/core/providers/op_kernel_type_control.h

```diff
@@ -238,6 +287,88 @@ struct EnabledTypes {
                                             ::onnxruntime::op_kernel_type_control::kAllOpSets, \
                                             ArgDirection, ArgIndex, __VA_ARGS__)
 
+/**
+ * Specifies a required set of types for a given Op kernel argument.
+ * Optional.
+ *
+ * Note: This should be called from the onnxruntime::op_kernel_type_control namespace.
+ *
+ * @param OpProvider The Op provider.
+ * @param OpDomain The Op domain.
+ * @param OpName The Op name.
+ * @param OpSet The opset that this set of required types applies to.
+ * @param ArgDirection Direction of the given Op kernel argument - Input or Output.
+ * @param ArgIndex Index of the given Op kernel argument.
+ * @param RequiredTypeList The types.
+ */
+#define ORT_SPECIFY_OP_KERNEL_ARG_REQUIRED_TYPE_LIST(                              \
+    OpProvider, OpDomain, OpName, OpSet, ArgDirection, ArgIndex, RequiredTypeList) \
+  class ORT_OP_KERNEL_TYPE_CTRL_INTERNAL_OP_TAG_CLASS_NAME(OpDomain, OpName);      \
+  class ORT_OP_KERNEL_TYPE_CTRL_INTERNAL_PROVIDER_TAG_CLASS_NAME(OpProvider);      \
+  template <>                                                                      \
+  struct ORT_OP_KERNEL_TYPE_CTRL_INTERNAL_REQUIRED_TYPES_HOLDER(                   \
+      OpProvider, OpDomain, OpName, OpSet, ArgDirection, ArgIndex) {               \
+    using types = RequiredTypeList;                                                \
+  };
+
+/**
+ * Specifies a required set of types for a given Op kernel argument that is valid for all opsets.
+ * Optional.
+ *
+ * Note: This should be called from the onnxruntime::op_kernel_type_control namespace.
+ *
+ * @param OpProvider The Op provider.
+ * @param OpDomain The Op domain.
+ * @param OpName The Op name.
+ * @param ArgDirection Direction of the given Op kernel argument - Input or Output.
+ * @param ArgIndex Index of the given Op kernel argument.
+ * @param RequiredTypeList The types.
+ */
+#define ORT_SPECIFY_OP_KERNEL_ARG_REQUIRED_TYPE_LIST_ALL_OPSETS(                                  \
+    OpProvider, OpDomain, OpName, ArgDirection, ArgIndex, RequiredTypeList)                       \
```

Comment:
Would a required type ever change between opsets?

Reply:
One example is Max, where the required int64_t is added in a later opset:
```
ORT_SPECIFY_OP_KERNEL_ARG_SUPPORTED_TYPES(kCpuExecutionProvider, kOnnxDomain, Max, 8, Input, 0, float, double);

ORT_SPECIFY_OP_KERNEL_ARG_SUPPORTED_TYPES(kCpuExecutionProvider, kOnnxDomain, Max, 12, Input, 0,
                                          float, double, MLFloat16, int32_t, uint32_t, int64_t, uint64_t);
ORT_SPECIFY_OP_KERNEL_ARG_REQUIRED_TYPES(kCpuExecutionProvider, kOnnxDomain, Max, 12, Input, 0,
                                         int64_t);
```